pyquil/_parser/parser.py

```diff
@@ -0,0 +1,554 @@
+        # new qubit arguments prefix the old ones.
+        modifier_qubits = []
+        for m in modifiers:
+            if m in ["CONTROLLED", "FORKED"]:
+                modifier_qubits.append(qubits[len(modifier_qubits)])
+
+        base_qubits = qubits[len(modifier_qubits) :]
+        forked_offset = len(params) >> modifiers.count("FORKED")
+        base_params = params[:forked_offset]
+
+        if name in QUANTUM_GATES:
+            if base_params:
+                gate = QUANTUM_GATES[name](*base_params, *base_qubits)
+            else:
+                gate = QUANTUM_GATES[name](*base_qubits)
+        else:
+            gate = Gate(name, base_params, base_qubits)
+
+        for modifier in modifiers[::-1]:
+            if modifier == "CONTROLLED":
+                gate.controlled(modifier_qubits.pop())
+            elif modifier == "DAGGER":
+                gate.dagger()
+            elif modifier == "FORKED":
+                gate.forked(modifier_qubits.pop(), params[forked_offset : (2 * forked_offset)])
+                forked_offset *= 2
+            else:
+                raise ValueError(f"Unsupported gate modifier {modifier}.")
+
+        return gate
+
+    modifiers = list
+    modifier = v_args(inline=True)(str)
+
+    @v_args(inline=True)
+    def frame(self, qubits, name):
+        f = Frame(qubits, name)
+        return f
+
+    @v_args(inline=True)
+    def pulse(self, nonblocking, frame, waveform):
+        p = Pulse(frame, waveform, nonblocking=bool(nonblocking))
+        return p
+
+    @v_args(inline=True)
+    def fence_some(self, qubits):
+        f = Fence(list(qubits))
+        return f
+
+    fence_all = v_args(inline=True)(FenceAll)
+
+    @v_args(inline=True)
+    def declare(self, name, memory_type, memory_size, *sharing):
+        shared, *offsets = sharing
+        d = Declare(
+            str(name),
+            memory_type=str(memory_type),
+            memory_size=int(memory_size) if memory_size else 1,
+            shared_region=str(shared) if shared else None,
+            offsets=offsets if shared else None,
+        )
+        return d
+
+    @v_args(inline=True)
+    def capture(self, nonblocking, frame, waveform, addr):
+        c = Capture(frame, waveform, addr, nonblocking=nonblocking)
+        return c
+
+    @v_args(inline=True)
+    def raw_capture(self, nonblocking, frame, expression, addr):
+        c = RawCapture(frame, expression, addr, nonblocking=nonblocking)
+        return c
+
+    @v_args(inline=True)
+    def addr(self, name):
+        return MemoryReference(str(name))
+
+    @v_args(inline=True)
+    def addr_subscript(self, name, subscript):
+        return MemoryReference(str(name), int(subscript))
+
+    @v_args(inline=True)
+    def offset_descriptor(self, offset, name):
+        return (int(offset), str(name))
+
+    @v_args(inline=True)
+    def delay_qubits(self, qubits, delay_amount=None):
+        # TODO(notmgsk): This is a very nasty hack. I can't quite get
```

Comment:
Yep, I ran into the same thing in the Rust parser. Brought that up [here](https://github.com/rigetti/pyquil/pull/1257#discussion_r505787345), @kilimanjaro and I must have discussed it on a chat.

I think this merits a change to the grammar, ie `DELAY 1 FOR 1`, but that will have to be sorted out in the spec first, leaving this "hack" in the meantime.

Reply:
Please update this comment to make it instructive on when to do what